### PR TITLE
Improve various signatures that shouldn't be `async def`, but currently are

### PR DESCRIPTION
**Pull request by AlexWaygood**
As discussed in #7475, the following methods are all "async def" functions at runtime, but should _not_ be "async def" functions in the stub:
- `AsyncIterator.__anext__`
- `AyncGenerator.__anext__`
- `AsyncGenerator.aclose`
- `AsyncGenerator.asend`
- `AsyncGenerator.athrow` 

`typing.AsyncIterator` and `typing.AsyncGenerator` represent abstract interfaces rather than concrete classes, and [PEP 525](https://peps.python.org/pep-0525/#support-for-asynchronous-iteration-protocol), which introduced the asynchronous iteration protocol, explicitly states that it is fine for these methods to return any awaitable. The current annotations, which state that these methods have to return _coroutines_, are therefore too restrictive, and cause false positives.

In addition to these changes:

- `AsyncGenerator.__anext__` is not abstract at runtime, and I don't see any reason why it should be in the stub either.
- `AsyncGenerator.__aiter__` is also not abstract at runtime, and not even defined on `AsyncGenerator` at runtime; it is simply inherited from `AsyncIterator`. I think we should do the same in the stub.
- `AsyncGenerator.aclose` is not abstract at runtime, and I don't see any reason why it should be in the stub.
- `_typeshed.SupportsAnext.__anext__` and `contextlib._SupportsAclose.aclose` were both erroneously made `async def` functions in #7107; this PR also reverts the changes _that_ PR made to _those_ classes.

This PR reverts the changes #7105 made to `typing.pyi`, but not the changes #7105 made to `types.pyi`. I think that makes sense, as the classes in `types` are concrete classes, whereas the classes in `typing` are abstract interfaces.

This PR adds several entries to the allowlist, but we can take them off again if https://github.com/python/mypy/pull/12343 is merged.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
The `boostedblob` error is related to this line [here](https://github.com/hauntsaninja/boostedblob/blob/47c1e0efaf3c15b35ace14b3c29694149555ae9f/boostedblob/boost.py#L547), and is because `asyncio.create_task` is annotated as requiring a _coroutine_ following #6779, and this PR would change the definition of `AsyncIterator.__anext__` such that it wouldn't necessarily be guaranteed to return a coroutine anymore.

**Comment by AlexWaygood** [comment]
Cc. @sobolevn, since this partially reverts your PR 🙂

**Comment by graingert** [comment]
Would SupportsAnext need fixing? https://github.com/python/typeshed/blob/master/stdlib/_typeshed/__init__.pyi#L36

**Comment by AlexWaygood** [comment]
> Would SupportsAnext need fixing? https://github.com/python/typeshed/blob/master/stdlib/_typeshed/__init__.pyi#L36

Good spot. `contextlib._SupportsAclose` as well.

**Comment by graingert** [comment]
anext is actually slightly odd as it returns whatever `__anext__` returns - eg an Awaitable or even not an awaitable!

```
Python 3.10.2 (main, Jan 15 2022, 18:02:07) [GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...     def __anext__(self):
...         return "hello"
... 
>>> anext(Foo())
'hello'
>>> 
```

**Comment by graingert** [comment]
so the type annotation for `builtins.anext` should look more like this:

```python
class SupportsAnyAnext(Protocol[_T_co]):
    def __anext__(self) -> _T_co: ...

class SupportsAwaitableAnext(Protcol[_T_co]):
    def __anext__(self) -> Awaitable[_T_co]: ...

@overload
def anext(__i: SupportsAnyAnext[_T]) -> _T: ...
@overload
async def anext(__i: SupportsAwaitableAnext[_T], default: _VT) -> _T | _VT: ...
```

so that anext doesn't incorrectly promote an `AsyncIterable.__next__` `Awaitable` into a `Coroutine`

**Comment by AlexWaygood** [comment]
> anext is actually slightly odd as it returns whatever `__anext__` returns - eg an Awaitable or even not an awaitable!

Sure, but I'd imagine that if `__anext__` returns something that isn't awaitable, it's almost certainly indicative of a programming error. So surely we'd be doing the developer a favour by raising an error if they try to call `anext` on an object that has an `__anext__` method that doesn't return an awaitable?

**Comment by graingert** [comment]
my point was more about preserving the return type of the `.__anext__` call: 

```python
_Awaitable_T_co = TypeVar("_Awaitable_T_co", bound=Awaitable)  # bound isn't strictly true - anext supports any return value

class SupportsAnext(Protocol[_Awaitable_T_co]):
    def __anext__(self) -> _Awaitable_T_co: ...

class SupportsAwaitableAnext(Protcol[_T_co]):
    def __anext__(self) -> Awaitable[_T_co]: ...

@overload
def anext(__i: SupportsAnyAnext[_T]) -> _T: ...
@overload
async def anext(__i: SupportsAwaitableAnext[_T], default: _VT) -> _T | _VT: ...
```

**Comment by AlexWaygood** [comment]
> my point was more about preserving the return type of the `.__anext__` call:

It's frying my brain somewhat, but I think I understand. Thanks for the explanation!

**Comment by AlexWaygood** [comment]
Better now @graingert?

File: stdlib/builtins.pyi
Comment (AlexWaygood) [reply stdlib/builtins.pyi]: Any suggestions for a good comment wording?

File: stdlib/builtins.pyi
Comment (graingert) [inline stdlib/builtins.pyi:1079]: here's a close python implementation of anext: https://github.com/python/cpython/blob/ea786a882b9ed4261eafabad6011bc7ef3b5bf94/Lib/test/test_asyncgen.py#L52-L80

```suggestion
    # anext is not, in fact, an async function. When default is not provided
    # anext is just a passthrough for obj.__anext__
    def anext(__i: _SupportsSynchronousAnext[_AwaitableT]) -> _AwaitableT: ...
```

maybe it's obvious now and doesn't need a comment :shrug:

Comment (graingert) [reply stdlib/builtins.pyi:1079]: to get an idea of what I'm trying to describe, here's a close python implementation of anext: https://github.com/python/cpython/blob/ea786a882b9ed4261eafabad6011bc7ef3b5bf94/Lib/test/test_asyncgen.py#L52-L80

**Review by graingert — APPROVED** [review]

File: stdlib/builtins.pyi
Comment (AlexWaygood) [reply stdlib/builtins.pyi:1081]: Oh, that's really useful! We should definitely include a link to that in the comment!

**Comment by AlexWaygood** [comment]
Thanks for the review @graingert! (I'll leave this open for a while to give other maintainers a chance to review.)

File: tests/stubtest_allowlists/py3_common.txt
Comment (graingert) [inline tests/stubtest_allowlists/py3_common.txt:249]: you probably want to link to 7491 as that's where most of the commentry is

**Comment by AlexWaygood** [comment]
(Planning to merge this in a day or so, unless any other maintainers want to review, since it's sort of blocking https://github.com/python/mypy/pull/12321)

**Pull request merged**
